WordPress-iOS-Shared/Core/OperationPools.swift

```diff
@@ -0,0 +1,286 @@
+        }
+    }
+}
+
+/// An OperationPool that transforms a NSURLRequest into a Value.
+///
+/// It uses `NSURLSession.downloadTaskWithRequest` to download the file for the
+/// given request, then uses a `Parser` to convert the downloaded file into the
+/// wanted `Value`.
+///
+public class NetworkDownloadPool<Value>: OperationPool<NSURLRequest, Value, NSError> {
+    public typealias Parser = NSURL -> Result<Value, NSError>
+
+    /// Constructs a new NetworkDownloadPool with the given parser.
+    ///
+    /// Optionally, you can specify the NSURLSession to use. By default, it uses
+    /// the shared session.
+    ///
+    /// - parameter parser: A function to be called to convert the downloaded
+    /// file to a `Value`. The `parser` function takes a file `NSURL` that
+    /// points to the location of the downloaded file, and returns a `Result`
+    /// with the processed `Value`, or a `NSError`.
+    ///
+    /// - parameter session: The NSURLSession to use for the network request. By
+    /// default it uses `NSURLSession.sharedSession()`.
+    ///
+    public init(parser: Parser, session: NSURLSession = NSURLSession.sharedSession()) {
+        super.init { request, completion in
+            let task = session.downloadTaskWithRequest(request) {
+                location, response, error in
+
+                let imageLocation = Result(value: location, error: error, fallbackError: OperationPoolError.noDataAndNoError.nserror)
+                let image = imageLocation.flatMap(parser)
+                completion(image)
+            }
+            task.resume()
+        }
+    }
+}
+
+/// An OperationPool to download images.
+///
+/// Builds on NetworkDownloadPool and transforms a `NSURLRequest` into a
+/// `UIImage`.
+///
+/// If the downloaded data can't be turned into an image, it results in a
+/// `OperationPoolError.invalidImageData` error.
+///
+public class ImageDownloadPool: NetworkDownloadPool<UIImage> {
+
+    /// Constructs a new ImageDownloadPool.
+    ///
+    /// Optionally, you can specify the NSURLSession to use. By default, it uses
+    /// the shared session.
+    ///
+    /// - parameter session: The NSURLSession to use for the network request. By
+    /// default it uses `NSURLSession.sharedSession()`.
+    ///
+    public init(session: NSURLSession = NSURLSession.sharedSession()) {
+        let parser: Parser = { url in
+            do {
+                let data = try NSData(contentsOfURL: url, options: .DataReadingUncached)
+                if let image = UIImage(data: data) {
+                    return .success(image)
+                } else {
+                    return .failure(OperationPoolError.invalidImageData.nserror)
+                }
+            } catch {
+                return .failure(error as NSError)
+            }
+        }
+        super.init(parser: parser, session: session)
+    }
+}
+
+/// Errors related to OperationPool.
+/// 
+public enum OperationPoolError: Int, ErrorType {
```

Comment:
I think I understand your point as: this conforms to `ErrorType` but it's being used as `NSError`.
And now I realize that `OperationPoolError` doesn't really need to conform to `ErrorType`, and it might be slightly less confusing if I remove that.

I believe the alternative to this would be to define a couple of separate `NSError` constants, but it feels slightly better to group errors inside an `enum`.

`ErrorType` is a bit weird in general.

We're kind of stuck with `NSError` since that's what we get from `NSURLSession`. An alternative would be to wrap those returned errors inside a `OperationPoolError` value, but I don't think it makes much sense: if there's an error it's probably going to come from `NSURLSession`, our custom errors are only there as sanity checks. So we don't want to be wrapping and unwrapping the most common error cases.

> why not just stick to Apple's new default of using exceptions for error handling? 

Because you can't `throw` an error asynchronously.